File: NLP_exercise.py
# Natural Language Processing

# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset
dataset = pd.read_csv('Restaurant_Reviews.tsv', delimiter = '\t', quoting = 3) # quoting 3 - means ignore double quotes

# Cleaning the texts
import re
import nltk #library for natural language processing
nltk.download('stopwords') #download the tool - stopwords list
from nltk.corpus import stopwords #import the stopwords
from nltk.stem.porter import PorterStemmer #import the stem tool - to keep the root of each word, get rid of its tense
corpus = []
for i in range(0, 1000):
    review = re.sub('[^a-zA-Z]', ' ', dataset['Review'][i]) #only keep letters in corpuss
    review = review.lower() #convert to lower case
    review = review.split() #convert string to list
    ps = PorterStemmer() #creat a PorterStemmer object
    review = [ps.stem(word) for word in review if not word in set(stopwords.words('english'))] #exlude word in the list of stopwords //set function make the algorithm much faster
    review = ' '.join(review) #join different words separated by space - convert list to string
    corpus.append(review)
    
# Creating the Bag of Words model
from sklearn.feature_extraction.text import CountVectorizer
cv = CountVectorizer(max_features = 1500) # keep 1500 most frequent words
X = cv.fit_transform(corpus).toarray() #create the sparse matrix .array() - to make a matrix
y = dataset.iloc[:, 1].values

# Splitting the dataset into the Training set and Test set
from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.20, random_state = 0)

# Fitting Naive Bayes to the Training set
from sklearn.naive_bayes import GaussianNB
classifier = GaussianNB()
classifier.fit(X_train, y_train)

# Predicting the Test set results
y_pred = classifier.predict(X_test)

# Making the Confusion Matrix
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_pred)
accuracy = (55+91)/200 #0.73
precision = 91/(91+42) #0.684
recall = 91/(91+12) #0.883
f1_score = 2 * precision * recall / (precision + recall) #0.77

# Fitting decision tree to the Training set
from sklearn.tree import DecisionTreeClassifier
classifier = DecisionTreeClassifier(criterion = 'entropy', random_state = 0)
classifier.fit(X_train, y_train)

# Predicting the Test set results
y_pred = classifier.predict(X_test)

# Making the Confusion Matrix
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_pred)
accuracy = (74+68)/200 #0.71
precision = 68/(68+23) #0.75
recall = 68/(68+35) #0.66
f1_score = 2 * precision * recall / (precision + recall) #0.70

# Fitting Random Forest classification to the Training set
from sklearn.ensemble import RandomForestClassifier
classifier = RandomForestClassifier(n_estimators = 500, criterion = 'entropy', random_state = 0)
classifier.fit(X_train, y_train)

# Predicting the Test set results
y_pred = classifier.predict(X_test)

# Making the Confusion Matrix
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_pred)
accuracy = (88+56)/200 #0.72
precision = 56/(56+9) #0.85
recall = 56/(56+47) #0.55
f1_score = 2 * precision * recall / (precision + recall) #0.667


# Fitting Logistic Regression to the Training set
from sklearn.linear_model import LogisticRegression
classifier = LogisticRegression(random_state = 0)
classifier.fit(X_train, y_train)

# Predicting the Test set results
y_pred = classifier.predict(X_test)

# Making the Confusion Matrix
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_pred)
accuracy = (76+66)/200 #0.71
precision = 66/(66+21) #0.758
recall = 66/(66+37) #0.64
f1_score = 2 * precision * recall / (precision + recall) #0.695

# Fitting KNN to the Training set
from sklearn.neighbors import KNeighborsClassifier
classifier = KNeighborsClassifier(n_neighbors = 5, metric = 'minkowski', p = 2) # use eclidean_distance
classifier.fit(X_train, y_train)

# Predicting the Test set results
y_pred = classifier.predict(X_test)

# Making the Confusion Matrix
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_pred)
accuracy = (74+48)/200 #0.61
precision = 48/(48+23) #0.68
recall = 48/(48+55) #0.47
f1_score = 2 * precision * recall / (precision + recall) #0.55

# Fitting SVM to the Training set
from sklearn.svm import SVC
classifier = SVC(kernel = 'linear', random_state = 0) # 'rbf', 'poly', 'sigmoid'
classifier.fit(X_train, y_train)

# Predicting the Test set results
y_pred = classifier.predict(X_test)

# Making the Confusion Matrix
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_pred)
accuracy = (70+74)/200 #0.72
precision = 70/(70+23) #0.75
recall = 70/(70+33) #0.68
f1_score = 2 * precision * recall / (precision + recall) #0.71


# Feature Scaling
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

# Fitting kernel-SVM/rbf to the Training set
from sklearn.svm import SVC
classifier = SVC(kernel = 'rbf', random_state = 0)
classifier.fit(X_train, y_train)

# Predicting the Test set results
y_pred = classifier.predict(X_test)

# Making the Confusion Matrix
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_pred)
accuracy = (83+64)/200 #0.735
precision = 83/(83+33) #0.716
recall = 83/(83+20) #0.806
f1_score = 2 * precision * recall / (precision + recall) #0.758

# Fitting kernel-SVM/sigmoid to the Training set
from sklearn.svm import SVC
classifier = SVC(kernel = 'sigmoid', random_state = 0)
classifier.fit(X_train, y_train)

# Predicting the Test set results
y_pred = classifier.predict(X_test)

# Making the Confusion Matrix
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_pred)
accuracy = (79+80)/200 #0.795
precision = 79/(79+17) #0.823
recall = 79/(79+24) #0.767
f1_score = 2 * precision * recall / (precision + recall) #0.794

File: NLP_exercise.R
# Natural Language Processing

# Importing the dataset
dataset_original = read.delim('Restaurant_Reviews.tsv', quote = '', stringsAsFactors = FALSE) #usually ignore quote in NLP

# Cleaning the texts
#install.packages('tm') - this package is used for NLP
#install.packages('SnowballC') - this package contains stopwords function
library(tm)
library(SnowballC)
corpus = VCorpus(VectorSource(dataset_original$Review)) #create the corpus matrix
corpus = tm_map(corpus, content_transformer(tolower)) #convert to lower case - as.character(corpus[[1]])
corpus = tm_map(corpus, removeNumbers) #remove all numbers
corpus = tm_map(corpus, removePunctuation) #remove punctuation
corpus = tm_map(corpus, removeWords, stopwords()) #remove unrelevant words
corpus = tm_map(corpus, stemDocument) #keep the root of a word/stemming
corpus = tm_map(corpus, stripWhitespace) #remove extra space

# Creating the Bag of Words model
dtm = DocumentTermMatrix(corpus) #create the sparse matrix
dtm = removeSparseTerms(dtm, 0.999) #keep 99.9% of the most frequent words in dtm (reduce almost 1000 words)
dataset = as.data.frame(as.matrix(dtm)) #transform the sparse matrix into a dataframe
dataset$Liked = dataset_original$Liked #add the dependency column to the dataset

# Encoding the target feature as factor
dataset$Liked = factor(dataset$Liked, levels = c(0, 1))

# Splitting the dataset into the Training set and Test set
# install.packages('caTools')
library(caTools)
set.seed(123)
split = sample.split(dataset$Liked, SplitRatio = 0.8)
training_set = subset(dataset, split == TRUE)
test_set = subset(dataset, split == FALSE)

# Fitting Random Forest Classification to the Training set
library(randomForest)
classifier = randomForest(x = training_set[-692],
                          y = training_set$Liked,
                          ntree = 10)

# Predicting the Test set results
y_pred = predict(classifier, newdata = test_set[-692])

# Making the Confusion Matrix
cm = table(test_set[, 692], y_pred)
accuracy = (82 + 77) / 200 #0.795
precision = 77 / (77 + 18) #0.811
recall = 77 / (77 + 23) #0.77
f1_score = 2 * precision * recall / (precision + recall) #0.79

# Fitting Logistic Regression to the Training set
classifier = glm(formula = Liked ~ .,
                 family = binomial,
                 data = training_set)

# Predicting the Test set results
prob_pred = predict(classifier, type = 'response', newdata = test_set[-692]) #remove the last column
y_pred = ifelse(prob_pred > 0.5, 1, 0)

# Making the Confusion Matrix
cm = table(test_set[, 692], y_pred)
accuracy = (49 + 61) / 200 #0.55
precision = 61 / (61 + 51) #0.545
recall = 61 / (61 + 39) #0.61
f1_score = 2 * precision * recall / (precision + recall) #0.575

# Fitting K-NN to the Training set and Predicting the Test set results
library(class)
y_pred = knn(train = training_set[, -692], 
             test = test_set[, -692],
             cl = training_set[, 692],
             k = 5)

# Making the Confusion Matrix
cm = table(test_set[, 692], y_pred)
accuracy = (75 + 54) / 200 #0.645
precision = 54 / (54 + 25) #0.684
recall = 54 / (54 + 46) #0.54
f1_score = 2 * precision * recall / (precision + recall) #0.603

# Fitting SVM to the Training set
library(e1071)
classifier = svm(formula = Liked ~ .,
                 data = training_set,
                 type = 'C-classification', # SVM could be used for classification and regression
                 kernel = 'linear') # 'linear' is the most basic SVM method


# Predicting the Test set results
y_pred = predict(classifier, newdata = test_set[-692])

# Making the Confusion Matrix
cm = table(test_set[, 692], y_pred)
accuracy = (78 + 81) / 200 #0.795
precision = 81 / (81 + 22) #0.786
recall = 81 / (81 + 19) #0.81
f1_score = 2 * precision * recall / (precision + recall) #0.798


# Fitting kernel SVM to the Training set
library(e1071)
classifier = svm(formula = Liked ~ .,
                 data = training_set,
                 type = 'C-classification', # SVM could be used for classification and regression
                 kernel = 'polynomial',
                 degree = 12) 


# Predicting the Test set results
y_pred = predict(classifier, newdata = test_set[-692])

# Making the Confusion Matrix
cm = table(test_set[, 692], y_pred)
accuracy = (71 + 82) / 200 #0.765
precision = 82 / (82 + 29) #0.739
recall = 82 / (82 + 18) #0.82
f1_score = 2 * precision * recall / (precision + recall) #0.777

# Fitting Decision Tree Classification to the Training set
library(rpart)
classifier = rpart(formula = Liked ~ .,
                   data = training_set)

# Predicting the Test set results
y_pred = predict(classifier, newdata = test_set[-692], type = 'class')

# Making the Confusion Matrix
cm = table(test_set[, 692], y_pred)
accuracy = (85 + 57) / 200 #0.71
precision = 57 / (57 + 15) #0.792
recall = 57 / (57 + 43) #0.57
f1_score = 2 * precision * recall / (precision + recall) #0.663

# Fitting CART Classification to the Training set
library(rpart)
classifier = rpart(formula = Liked ~ .,
                   data = training_set,
                   method = 'class',
                   control = rpart.control(cp = 0.0, minsplit = 2))

# Predicting the Test set results
y_pred = predict(classifier, newdata = test_set[-692], type = 'class')

# Making the Confusion Matrix
cm = table(test_set[, 692], y_pred)
accuracy = (84 + 64) / 200 #0.74
precision = 64 / (64 + 16) #0.8
recall = 64 / (64 + 36) #0.64
f1_score = 2 * precision * recall / (precision + recall) #0.711

# Fitting Maximum Entropy Classification to the Training set
#install.packages('maxent')
library(maxent)
classifier = maxent(feature_matrix = training_set[-692],
                    code_vector = training_set$Liked)

# Predicting the Test set results
y_pred = predict(classifier, test_set[-692])
y_pred = y_pred[,1]

# Making the Confusion Matrix
cm = table(test_set[, 692], y_pred)
accuracy = (83 + 69) / 200 #0.76
precision = 83 / (83 + 31) #0.728
recall = 83 / (83 + 17) #0.83
f1_score = 2 * precision * recall / (precision + recall) #0.776

# Fitting C5.0 Classification to the Training set
#install.packages('C50')
library(C50)
classifier = C5.0(x = training_set[-692],
                  y = training_set$Liked)

# Predicting the Test set results
y_pred = predict(classifier, newdata = test_set[-692], type = 'class')

# Making the Confusion Matrix
cm = table(test_set[, 692], y_pred)
accuracy = (91 + 37) / 200 #0.64
precision = 37 / (37 + 9) #0.804
recall = 37 / (37 + 63) #0.37
f1_score = 2 * precision * recall / (precision + recall) #0.506


# Fitting Naive Bayes to the Training set
library(e1071)
classifier = naiveBayes(x = training_set[-692],
                        y = training_set$Liked)

# Predicting the Test set results
y_pred = predict(classifier, newdata = test_set[-692])

# Making the Confusion Matrix
cm = table(test_set[, 692], y_pred)
accuracy = (96 + 5) / 200 #0.505
precision = 96 / (96 + 95) #0.503
recall = 96 / (96 + 4) #0.96
f1_score = 2 * precision * recall / (precision + recall) #0.659